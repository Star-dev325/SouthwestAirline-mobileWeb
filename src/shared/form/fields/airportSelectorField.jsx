// @flow

import _ from 'lodash';
import React from 'react';
import cx from 'classnames';
import Icon from 'src/shared/components/icon';
import AirportList from 'src/airports/components/airportList';
import withField from 'src/shared/form/enhancers/withField';
import ClickableDiv from 'src/shared/components/clickableDiv';
import FullScreenModal from 'src/shared/components/fullScreenModal/fullScreenModal';
import { getAirportFromCode, getAirportFromAirportGroupShortDisplayName } from 'src/airports/helpers/airportsHelpers';
import {
  showFullScreenModal,
  hideFullScreenModal
} from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';

import type { FieldProps } from 'src/shared/form/flow-typed/form.types';
import type { AirportType, MultiSelectGroup } from 'src/shared/flow-typed/shared.types';

type Airport = {
  code: string,
  airportName: string,
  cityState: string,
  airportGroupShortDisplayName: string,
  airportGroupSubtitle: string
};

type Props = {
  allAirports: Array<AirportType>,
  clickableClassName: string,
  containerClassName?: string,
  dataForE2E?: string,
  description?: string,
  disabled: boolean,
  horizontalLayout?: boolean,
  iconRight: boolean,
  iconType: string,
  isMultiSelectGroupEnabled: boolean,
  isReaccomCoTerminalEligible?: boolean,
  isWebView?: boolean,
  modalId: string,
  multiSelectGroup: MultiSelectGroup,
  onClick?: () => void,
  placeholder: string,
  recentlySearched: Array<AirportType>,
  updateSelectedAirportInfoFn: (airportInfo: *) => void,
  usingNativeStyle?: boolean
} & FieldProps;

class AirportSelectorField extends React.Component<Props, *> {
  static defaultProps = {
    disabled: false,
    iconRight: false,
    clickableClassName: '',
    isMultiSelectGroupEnabled: false,
    multiSelectGroup: {}
  };

  _onAirportSelect = (airport: Airport, isCurrentLocation: boolean) => {
    const { name, modalId, onChange, updateSelectedAirportInfoFn } = this.props;

    const airportInfo = {
      [name]: {
        isCurrentLocation
      }
    };

    hideFullScreenModal(modalId);

    updateSelectedAirportInfoFn(airportInfo);
    onChange(airport.code);
  };

  _shouldReduceFontSize(code) {
    const MAX_WORD_CHAR = 8;

    return code.split(' ').find((word) => word.length > MAX_WORD_CHAR);
  }

  _renderFormattedAirport(airport: Airport) {
    const { name, isMultiSelectGroupEnabled, multiSelectGroup, containerClassName } = this.props;
    const { airportName, cityState, airportGroupShortDisplayName, airportGroupSubtitle } = airport;
    const { disabled, usingNativeStyle } = this.props;
    let code = airport.code;
    let labelText = `${airportName}, ${cityState}`;

    if (isMultiSelectGroupEnabled && multiSelectGroup[name] && multiSelectGroup[name].length > 1) {
      code = airportGroupShortDisplayName;
      labelText = airportGroupSubtitle;
    }

    if (usingNativeStyle) {
      return (
        <div className={cx("form-field--text-container", containerClassName)}>
          <div
            className={cx('form-field--text', {
              'form-field--long-text': this._shouldReduceFontSize(code)
            })}
          >
            {code}
          </div>
          <div className="form-field--description">{labelText}</div>
        </div>
      );
    } else {
      return (
        <div className="flex flex-cross-center">
          <div
            className={cx('airport-selector--formatted-airport', {
              'airport-selector--formatted-airport_disabled': disabled
            })}
          >
            {code}
          </div>
          <span className="airport-selector--airport-description">{labelText}</span>
        </div>
      );
    }
  }

  _renderPlaceholder = ({ placeholder, description }) => {
    const { usingNativeStyle, MWEB_HOMEPAGE_REDESIGN, containerClassName } = this.props;

    const placeholderClass = usingNativeStyle ? 'form-field--placeholder' : 'airport-selector--placeholder';
    const descriptionClass = usingNativeStyle ? 'form-field--description' : 'small';

    return (
      <div
        className={cx(
          {
            'form-field--text-container': usingNativeStyle,
            'airport-selector--placeholder-container py3': !usingNativeStyle
          },
          containerClassName
        )}
      >
        <div className={placeholderClass}>{placeholder}</div>
        {description && (
          <div className={descriptionClass}>
            {MWEB_HOMEPAGE_REDESIGN && (
              <Icon type="exclamation-circle" className="form-field--description-error-icon" />
            )}
            {description}
          </div>
        )}
      </div>
    );
  };

  _renderAirportLabel = () => {
    const { value, placeholder, description, allAirports, isMultiSelectGroupEnabled, multiSelectGroup, name } =
      this.props;

    if (value && allAirports !== null && allAirports.length > 0) {
      let airportData = getAirportFromCode(allAirports, value);

      if (isMultiSelectGroupEnabled && multiSelectGroup[name] && multiSelectGroup[name].length > 1) {
        airportData = getAirportFromAirportGroupShortDisplayName(allAirports, multiSelectGroup[name]);
      }

      return this._renderFormattedAirport(airportData);
    }

    return this._renderPlaceholder({ placeholder, description });
  };

  _handleSelectorClick = () => {
    const { onClick = _.noop, modalId, disabled } = this.props;

    onClick();
    !disabled && showFullScreenModal(modalId);
  };

  _renderClickableDiv = () => {
    const { iconType, iconRight, clickableClassName } = this.props;

    return (
      <ClickableDiv
        onClick={this._handleSelectorClick}
        iconType={iconType}
        iconRight={iconRight}
        className={clickableClassName}
        childClassName="ellipsis"
      >
        {this._renderAirportLabel()}
      </ClickableDiv>
    );
  };

  _renderNativeClickableDiv = () => {
    const { iconType, clickableClassName, horizontalLayout } = this.props;

    return (
      <div onClick={this._handleSelectorClick} className={cx(clickableClassName, 'form-field--container')}>
        {this._renderNativeIcon(iconType)}
        {this._renderAirportLabel()}
        {!horizontalLayout && (
          <div className="form-field--error-icon">
            <Icon type="exclamation-circle" />
          </div>
        )}
      </div>
    );
  };

  _renderNativeIcon = (iconType) => {
    if (iconType) {
      return (
        <div className="form-field--icon">
          <Icon type={iconType} />
        </div>
      );
    }

    return null;
  };

  _renderSelectorField = () => {
    const { usingNativeStyle } = this.props;

    if (usingNativeStyle) {
      return this._renderNativeClickableDiv();
    } else {
      return this._renderClickableDiv();
    }
  };

  _renderAirportList = (modalId?: string) => {
    const {
      allAirports,
      isMultiSelectGroupEnabled,
      isReaccomCoTerminalEligible,
      isWebView,
      multiSelectGroup,
      recentlySearched
    } = this.props;

    return (
      <AirportList
        allAirports={allAirports}
        disableInternationals={false}
        isMultiSelectGroupEnabled={isMultiSelectGroupEnabled}
        isReaccomCoTerminalEligible={isReaccomCoTerminalEligible}
        isWebView={isWebView}
        modalId={modalId}
        multiSelectGroup={multiSelectGroup}
        onAirportSelect={this._onAirportSelect}
        onCancel={() => hideFullScreenModal(modalId)}
        recentlySearched={recentlySearched}
        title="Select Airport"
      />
    );
  };

  render() {
    const { dataForE2E, modalId, usingNativeStyle, horizontalLayout } = this.props;

    return (
      <div
        className={`airport-selector${
          usingNativeStyle && horizontalLayout ? ' airport-selector_native-horizontal' : ''
        }`}
        data-qa={dataForE2E}
      >
        {this._renderSelectorField()}
        <FullScreenModal id={modalId}>{this._renderAirportList(modalId)}</FullScreenModal>
      </div>
    );
  }
}

export default withField()(AirportSelectorField);
