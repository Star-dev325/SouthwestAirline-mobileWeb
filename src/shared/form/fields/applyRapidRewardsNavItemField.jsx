// @flow
import i18n from '@swa-ui/locale';
import cx from 'classnames';
import React from 'react';
import NavItemLink from 'src/shared/components/navItemLink';
import withField from 'src/shared/form/enhancers/withField';
import type { FieldProps } from 'src/shared/form/flow-typed/form.types';

type Props = FieldProps & {
  onNavItemClick: () => void,
  rapidRewardsApplied: boolean
};

const ApplyRapidRewardsNavItemField = (props: Props) => {
  const { onNavItemClick, rapidRewardsApplied } = props;

  return (
    <div data-qa="review-form--apply-rapid-rewards-nav-item">
      <NavItemLink
        onClick={onNavItemClick}
        className={`nav-item-field ${cx({ 'no-points-selected': !rapidRewardsApplied })}`}
        icon={'keyboard-arrow-right'}
        iconClassName={'nav-item-link--icon'}
      >
        <span>{rapidRewardsApplied ? i18n('SPLIT_PAY_PAGE__CASH_POINTS_APPLIED_TEXT') : i18n('SPLIT_PAY_FORM__SELECT_TEXT') }</span>
      </NavItemLink>
    </div>
  );
};

export default withField()(ApplyRapidRewardsNavItemField);
