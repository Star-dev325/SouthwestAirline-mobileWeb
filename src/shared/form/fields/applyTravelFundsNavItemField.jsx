// @flow

import React from 'react';
import cx from 'classnames';

import NavItemLink from 'src/shared/components/navItemLink';
import withField from 'src/shared/form/enhancers/withField';

import type { FieldProps } from 'src/shared/form/flow-typed/form.types';

type Props = FieldProps & {
  onNavItemClick: () => void,
  travelFundsApplied: boolean
};

const ApplyTravelFundsNavItemField = (props: Props) => {
  const { onNavItemClick, travelFundsApplied } = props;

  return (
    <div data-qa="review-form--apply-travel-funds-nav-item">
      <NavItemLink
        onClick={onNavItemClick}
        className={`nav-item-field ${cx({ 'no-funds-selected': !travelFundsApplied })}`}
        icon={'keyboard-arrow-right'}
        iconClassName={'nav-item-link--icon'}
      >
        <span>{travelFundsApplied ? 'Funds Applied' : 'Select (optional)'}</span>
      </NavItemLink>
    </div>
  );
};

export default withField()(ApplyTravelFundsNavItemField);
