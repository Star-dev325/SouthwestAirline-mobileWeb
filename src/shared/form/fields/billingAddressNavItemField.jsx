// @flow

import React from 'react';

import NavItemLink from 'src/shared/components/navItemLink';
import withField from 'src/shared/form/enhancers/withField';

import type { FieldProps } from 'src/shared/form/flow-typed/form.types';

type Props = FieldProps & {
  onNavItemClick: () => void,
  billingAddressComplete: boolean
};

const BillingAddressNavItemField = (props: Props) => {
  const { onNavItemClick, billingAddressComplete } = props;

  return (
    <div data-qa="review-form--billing-address-nav-item">
      {billingAddressComplete && (
        <NavItemLink
          onClick={onNavItemClick}
          className={'nav-item-field'}
          icon={'keyboard-arrow-right'}
          iconClassName={'nav-item-link--icon'}
        >
          {<span className="pdkblue">Complete</span>}
        </NavItemLink>
      )}
      {!billingAddressComplete && (
        <NavItemLink
          onClick={onNavItemClick}
          className={'nav-item-field'}
          icon={'exclamation-circle warning'}
          iconClassName={'nav-item-link--bang-icon'}
        >
          {<span className="gray3">Required</span>}
        </NavItemLink>
      )}
    </div>
  );
};

export default withField()(BillingAddressNavItemField);
