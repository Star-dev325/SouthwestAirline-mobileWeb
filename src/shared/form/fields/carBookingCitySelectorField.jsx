// @flow
import React from 'react';
import _ from 'lodash';
import Icon from 'src/shared/components/icon';
import withField from 'src/shared/form/enhancers/withField';
import CarLocations from 'src/carBooking/components/carLocations';
import ClickableDiv from 'src/shared/components/clickableDiv';
import FullScreenModal from 'src/shared/components/fullScreenModal/fullScreenModal';
import {
  showFullScreenModal,
  hideFullScreenModal
} from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';
import {
  CAR_BOOKING_PICKUP_CITY_MODAL_ID,
  CAR_BOOKING_DROPOFF_CITY_MODAL_ID
} from 'src/carBooking/constants/carBookingConstants';
import i18n from '@swa-ui/locale';

import type { CarLocationResponseType } from 'src/carBooking/flow-typed/carBooking.types';

type PickUpAndDropOffCityType = {
  pickUp: string,
  dropOff: string
};

type Props = {
  carLocations: Array<CarLocationResponseType>,
  isWebView?: boolean,
  onChange: (*) => void,
  retrieveCarLocationsFn: () => void,
  value: PickUpAndDropOffCityType
};

export class CarBookingCitySelectorField extends React.Component<Props> {
  _onPickUpAirportSelect = (location: CarLocationResponseType) => {
    const {
      onChange,
      value: { dropOff }
    } = this.props;
    const pickUp = location.airport.code;
    const newDropOff = _.isEmpty(dropOff) ? location.airport.code : dropOff;
    const newValue = {
      pickUp,
      dropOff: newDropOff
    };

    onChange(newValue);

    hideFullScreenModal(CAR_BOOKING_PICKUP_CITY_MODAL_ID);
  };

  _onDropOffAirportSelect = (location: CarLocationResponseType) => {
    const {
      onChange,
      value: { pickUp }
    } = this.props;
    const newPickUp = _.isEmpty(pickUp) ? location.airport.code : pickUp;
    const dropOff = location.airport.code;
    const newValue = {
      pickUp: newPickUp,
      dropOff
    };

    onChange(newValue);

    hideFullScreenModal(CAR_BOOKING_DROPOFF_CITY_MODAL_ID);
  };

  _openSelectCarLocationModal = (modalId: string) => {
    const { carLocations, retrieveCarLocationsFn } = this.props;

    if (_.isEmpty(carLocations)) {
      retrieveCarLocationsFn && retrieveCarLocationsFn();
    }
    showFullScreenModal(modalId);
  };

  _getCarLocationFromCode = (carLocationCode: ?string) =>
    _.find(this.props.carLocations, (location) => location.airport.code === carLocationCode);

  _renderCity = (fieldName: string, carLocationCode: string, placeholder: string, modalId: string) => (
    <div className="car-city-selector center" data-qa={`car-${_.kebabCase(fieldName || '')}`}>
      <div className="center">
        <ClickableDiv
          childClassName="fullwidth"
          iconType="car"
          onClick={() => this._openSelectCarLocationModal(modalId)}
        >
          {this._renderCarCityLabel(carLocationCode, placeholder)}
        </ClickableDiv>
      </div>
    </div>
  );

  _renderCarCityLabel(carLocationCode: ?string, placeholder: ?string) {
    const carLocation = _.isEmpty(carLocationCode) ? {} : this._getCarLocationFromCode(carLocationCode);
    const formatCarCity = (location = {}) => {
      const { airport, city, state } = location;

      return !!airport && !!city ? `${city}, ${state}` : '';
    };

    if (carLocationCode && carLocation) {
      return (
        <div>
          <span className={'bold pdkblue xxxlarge block'}>{_.get(carLocation, 'airport.code')}</span>
          <span className={'gray4 large mt3 block word-wrap'}>{formatCarCity(carLocation)}</span>
        </div>
      );
    } else {
      return (
        <div>
          <span className={'gray4 xxxlarge block'}>{placeholder}</span>
          <span className={'gray4 large mt3 block'}>{i18n('CAR_BOOKING__SELECT')}</span>
        </div>
      );
    }
  }

  render() {
    const {
      value: { pickUp, dropOff },
      carLocations,
      isWebView
    } = this.props;

    return (
      <div className="car-booking-search-form--triptych flex flex-main-center flex-cross-start p5 bdgray2">
        <div className="car-booking-search-form--triptych-side">
          {this._renderCity('pickUp', pickUp, 'Pick-up', CAR_BOOKING_PICKUP_CITY_MODAL_ID)}
        </div>
        <div className="car-booking-search-form--triptych-center flex-item-center sltblue xxxlarge m1">
          {<Icon type="car" />}
        </div>
        <div className="car-booking-search-form--triptych-side">
          {this._renderCity('dropOff', dropOff, 'Return', CAR_BOOKING_DROPOFF_CITY_MODAL_ID)}
        </div>
        <FullScreenModal id={CAR_BOOKING_PICKUP_CITY_MODAL_ID}>
          <CarLocations
            onCancel={() => hideFullScreenModal(CAR_BOOKING_PICKUP_CITY_MODAL_ID)}
            onAirportSelect={this._onPickUpAirportSelect}
            title="Select City"
            carLocations={carLocations}
            isWebView={isWebView}
          />
        </FullScreenModal>
        <FullScreenModal id={CAR_BOOKING_DROPOFF_CITY_MODAL_ID}>
          <CarLocations
            onCancel={() => hideFullScreenModal(CAR_BOOKING_DROPOFF_CITY_MODAL_ID)}
            onAirportSelect={this._onDropOffAirportSelect}
            title="Select City"
            carLocations={carLocations}
            isWebView={isWebView}
          />
        </FullScreenModal>
      </div>
    );
  }
}

export default withField()(CarBookingCitySelectorField);
