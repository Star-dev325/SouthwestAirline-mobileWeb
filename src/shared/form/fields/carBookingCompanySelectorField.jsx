// @flow
import i18n from '@swa-ui/locale';
import _ from 'lodash';
import React from 'react';
import CarBookingCompanyList from 'src/carBooking/components/carBookingCompanyList';
import { CAR_BOOKING_COMPANY_MODAL_ID } from 'src/carBooking/constants/carBookingConstants';
import {
  transformToCarCompanyList,
  transformToCarCompanyViewValue
} from 'src/carBooking/transformers/carVendorTransformer';
import FullScreenModal from 'src/shared/components/fullScreenModal/fullScreenModal';
import {
  hideFullScreenModal,
  showFullScreenModal
} from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';
import withField from 'src/shared/form/enhancers/withField';

import type { CarCompanyType, CarVendorType, CompanyVendorType } from 'src/carBooking/flow-typed/carBooking.types';

type Props = {
  carVendors: Array<CarVendorType>,
  isWebView?: boolean,
  name: string,
  onChange: (*) => void,
  retrieveCarVendorsFn: () => void,
  value: string | Array<CompanyVendorType>
};

type State = {
  carCompanyArray: Array<Array<CarCompanyType>>,
  viewValue: string
};

class CarBookingCompanySelectorField extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const { carVendors, value } = this.props;
    const viewValue = _.isEmpty(carVendors) ? 'Shop all' : transformToCarCompanyViewValue(carVendors, value);

    this.state = {
      viewValue,
      carCompanyArray: transformToCarCompanyList(carVendors, value)
    };
  }

  areCarVendorListsEqual = (vendors1, vendors2) =>
    vendors1.length === vendors2.length &&
    vendors1.filter((vendor, index) => vendor.vendorId === vendors2[index].vendorId).length === vendors1.length;

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    const { carVendors, value } = nextProps;

    const viewValue = transformToCarCompanyViewValue(carVendors, value);

    this.setState({
      carCompanyArray: this.areCarVendorListsEqual(nextProps.carVendors, this.props.carVendors)
        ? this.state.carCompanyArray
        : transformToCarCompanyList(carVendors, value),
      viewValue
    });
  }

  _onCarCompanyGroupSelectFn = (carCompany: CarCompanyType, groupIndex: number) => {
    const { carCompanyArray } = this.state;
    const { name, isSelected } = carCompany;

    if (name === i18n('CAR_BOOKING__CAR_VENDOR__SHOP_ALL_OPTION')) {
      _.forEach(carCompanyArray, (carVendorGroup) => {
        this._selectCarCompanyGroup(carVendorGroup, isSelected);
      });
    } else if (
      name === i18n('CAR_BOOKING__CAR_VENDOR__RAPID_REWARDS_PARTNERS') ||
      name === i18n('CAR_BOOKING__CAR_VENDOR__SHOP_ALL_OTHERS_OPTION')
    ) {
      this._selectCarCompanyGroup(carCompanyArray[groupIndex], isSelected);
    }

    const carCompanyGroup = carCompanyArray[groupIndex];
    const notAllChildOptionSelected = _.chain(carCompanyGroup)
      .slice(1, carCompanyGroup.length)
      .map('isSelected')
      .includes(false)
      .value();

    notAllChildOptionSelected
      ? this.setState(_.merge(carCompanyArray[groupIndex][0], { isSelected: false }))
      : this.setState(_.merge(carCompanyArray[groupIndex][0], { isSelected: true }));

    this.setState(
      _.merge(carCompanyArray[0][0], {
        isSelected: carCompanyArray[1][0].isSelected && carCompanyArray[2][0].isSelected
      })
    );
  };

  _onCarCompanySelectFn = (carCompany: CarCompanyType, groupIndex: number) => {
    const { carCompanyArray } = this.state;
    const carVendorGroup = carCompanyArray[groupIndex];
    const index = _.findIndex(carVendorGroup, (company) => company.name === carCompany.name);

    carVendorGroup[index] = carCompany;

    if (carCompany.isSelected) {
      const notAllChildOptionSelected = _.chain(carVendorGroup)
        .slice(1, carVendorGroup.length)
        .map('isSelected')
        .includes(false)
        .value();

      carCompanyArray[groupIndex][0].isSelected = !notAllChildOptionSelected;
      carCompanyArray[0][0].isSelected = carCompanyArray[1][0].isSelected && carCompanyArray[2][0].isSelected;
    } else {
      carCompanyArray[0][0].isSelected = false;
      carCompanyArray[groupIndex][0].isSelected = false;
    }
    this.setState({ carCompanyArray });
  };

  _onDoneClickFn = () => {
    const { carCompanyArray } = this.state;
    const { onChange } = this.props;

    let selectedVendors;
    const rapidRewordsPartnerSelected: Array<CarCompanyType> = [];
    const rapidRewordsPartnerUnselected: Array<CarCompanyType> = [];
    const allOthersSelected: Array<CarCompanyType> = [];
    const allOthersUnselected: Array<CarCompanyType> = [];

    _.forEach(carCompanyArray[1].slice(1, carCompanyArray[1].length), (carVendor) => {
      carVendor.isSelected
        ? rapidRewordsPartnerSelected.push(carVendor)
        : rapidRewordsPartnerUnselected.push(carVendor);
    });
    _.forEach(carCompanyArray[2].slice(1, carCompanyArray[2].length), (carVendor) => {
      carVendor.isSelected ? allOthersSelected.push(carVendor) : allOthersUnselected.push(carVendor);
    });
    const shopAllSelect = carCompanyArray[0][0].isSelected;
    const nothingSelect = rapidRewordsPartnerSelected.length + allOthersSelected.length === 0;

    if (shopAllSelect || nothingSelect) {
      selectedVendors = i18n('CAR_BOOKING__CAR_VENDOR__SHOP_ALL_HINT');
    } else {
      selectedVendors = this._getVendorIdAndName(rapidRewordsPartnerSelected.concat(allOthersSelected));
    }

    onChange(selectedVendors);
    hideFullScreenModal(CAR_BOOKING_COMPANY_MODAL_ID);
  };

  _selectCarCompanyGroup = (carVendors: Array<CarCompanyType>, isSelected: boolean) => {
    _.forEach(carVendors, (carVendor) => {
      carVendor.isSelected = isSelected;
    });
  };

  _getVendorIdAndName = (vendorList: Array<CarCompanyType>) =>
    _.map(vendorList, (vendor) => ({
      vendorId: vendor.vendorId,
      vendorName: vendor.name
    }));

  _openSelectCarVendorPage = () => {
    const { carVendors, retrieveCarVendorsFn, value } = this.props;

    if (_.isEmpty(carVendors)) {
      retrieveCarVendorsFn && retrieveCarVendorsFn();
    }
    const carCompanyArray = transformToCarCompanyList(carVendors, value);

    this.setState({ carCompanyArray });
    showFullScreenModal(CAR_BOOKING_COMPANY_MODAL_ID);
  };

  render() {
    const { viewValue, carCompanyArray } = this.state;
    const { isWebView } = this.props;

    return (
      <div>
        <div
          className="flex flex-main-between flex-cross-center px5"
          onClick={this._openSelectCarVendorPage}
          data-qa="car-companies"
        >
          <label className="xlarge">{i18n('CAR_BOOKING__CAR_VENDOR_FORM__TITLE')}</label>
          <div data-qa="car-booking-company" className="gray4">
            {viewValue}
          </div>
        </div>
        <FullScreenModal id={CAR_BOOKING_COMPANY_MODAL_ID}>
          <CarBookingCompanyList
            carVendors={carCompanyArray}
            isWebView={isWebView}
            onCancel={() => hideFullScreenModal(CAR_BOOKING_COMPANY_MODAL_ID)}
            onCarCompanyGroupSelectFn={this._onCarCompanyGroupSelectFn}
            onCarCompanySelectDoneFn={this._onDoneClickFn}
            onCarCompanySelectFn={this._onCarCompanySelectFn}
          />
        </FullScreenModal>
      </div>
    );
  }
}

export default withField()(CarBookingCompanySelectorField);
