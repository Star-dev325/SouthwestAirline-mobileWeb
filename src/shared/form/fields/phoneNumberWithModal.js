// @flow
import _ from 'lodash';
import PhoneNumberFields from 'src/shared/form/fields/phoneNumberFields';
import React from 'react';
import PropTypes from 'prop-types';
import {
  hideFullScreenModal,
  showFullScreenModal
} from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';
import FullScreenModal from 'src/shared/components/fullScreenModal/fullScreenModal';
import PhoneCountryCodeList from 'src/shared/components/phoneCountryCodeList';
import countryCodes from 'src/shared/constants/countryCode';

type PropsWithModal = {
  nameForPhoneCountryCode: string,
  nameForPhoneNumber: string,
  isISOCountryCode?: boolean,
  className?: string
};

export class PhoneNumberWithModal extends React.Component<PropsWithModal> {
  static contextTypes = {
    form: PropTypes.object
  };

  render() {
    const { nameForPhoneCountryCode, nameForPhoneNumber, isISOCountryCode } = this.props;
    const COUNTRY_CODE_MODAL_ID = _.toUpper(_.snakeCase(nameForPhoneCountryCode));

    const updateCountryCode = ({ countryCode }: { countryCode: string }) => {
      const { onChange } = this.context.form;
      const countryCodeValue = isISOCountryCode ? countryCode : `${countryCodes[countryCode]}`;

      hideFullScreenModal(COUNTRY_CODE_MODAL_ID);
      onChange && onChange(nameForPhoneCountryCode, countryCodeValue);
      onChange && onChange(nameForPhoneNumber, '');
    };

    return (
      <div>
        <PhoneNumberFields
          names={[nameForPhoneCountryCode, nameForPhoneNumber]}
          {...this.props}
          onLabelClick={() => showFullScreenModal(COUNTRY_CODE_MODAL_ID)}
        />
        <FullScreenModal id={COUNTRY_CODE_MODAL_ID}>
          <PhoneCountryCodeList
            onCountryCodeSelect={updateCountryCode}
            onCancel={() => hideFullScreenModal(COUNTRY_CODE_MODAL_ID)}
          />
        </FullScreenModal>
      </div>
    );
  }
}

export default PhoneNumberWithModal;
