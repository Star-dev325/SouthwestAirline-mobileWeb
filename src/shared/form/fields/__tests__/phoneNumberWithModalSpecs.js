import React from 'react';
import { PhoneNumberWithModal } from 'src/shared/form/fields/phoneNumberWithModal';
import { shallow } from 'enzyme';
import * as FullScreenModalHelper from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';
import { sandbox } from 'sinon';

const sinon = sandbox.create();

describe('Phone Number With Modal', () => {
  let onChangeStub;
  let hideFullScreenModalStub;
  let showFullScreenModalStub;

  beforeEach(() => {
    onChangeStub = sinon.stub();
    showFullScreenModalStub = sinon.stub(FullScreenModalHelper, 'showFullScreenModal');
    hideFullScreenModalStub = sinon.stub(FullScreenModalHelper, 'hideFullScreenModal');
  });

  afterEach(() => {
    sinon.restore();
  });

  it('should render default props', () => {
    expect(createComponent()).toMatchSnapshot();
  });

  it('should render non-iso props', () => {
    expect(
      createComponent({
        nameForPhoneCountryCode: 'testCaseCountryCode',
        nameForPhoneNumber: 'testCasePhoneNumber',
        isISOCountryCode: false,
        className: 'non-iso-codes'
      })
    ).toMatchSnapshot();
  });

  it('should open country modal on phone label click', () => {
    const component = createComponent();

    component.find('WithFields(PhoneNumberFields)').props().onLabelClick();

    expect(showFullScreenModalStub).to.have.been.calledWith('COUNTRY_CODE');
  });

  it('should reset phone number on country selection', () => {
    const component = createComponent();

    component.find('PhoneCountryCodeList').props().onCountryCodeSelect({ countryCode: 'NZ' });

    expect(onChangeStub.getCall(0).args).to.have.deep.equal(['countryCode', 'NZ']);
    expect(onChangeStub.getCall(1).args).to.have.deep.equal(['phoneNumber', '']);
  });

  it('should reset phone number on country selection for non-iso', () => {
    const component = createComponent({
      nameForPhoneCountryCode: 'testCaseCountryCode',
      nameForPhoneNumber: 'testCasePhoneNumber',
      isISOCountryCode: false
    });

    component.find('PhoneCountryCodeList').props().onCountryCodeSelect({ countryCode: 'NZ' });

    expect(onChangeStub.getCall(0).args).to.have.deep.equal(['testCaseCountryCode', '64']);
    expect(onChangeStub.getCall(1).args).to.have.deep.equal(['testCasePhoneNumber', '']);
  });

  it('should close country modal on cancel', () => {
    const component = createComponent();

    component.find('PhoneCountryCodeList').props().onCancel();

    expect(hideFullScreenModalStub).to.have.been.calledWith('COUNTRY_CODE');
  });

  function createComponent(props) {
    const context = {
      form: {
        onChange: onChangeStub
      }
    };
    const defaultProps = {
      nameForPhoneCountryCode: 'countryCode',
      nameForPhoneNumber: 'phoneNumber',
      isISOCountryCode: true
    };

    return shallow(
      <PhoneNumberWithModal
        {...{
          ...defaultProps,
          ...props
        }}
      />,
      { context }
    );
  }
});
