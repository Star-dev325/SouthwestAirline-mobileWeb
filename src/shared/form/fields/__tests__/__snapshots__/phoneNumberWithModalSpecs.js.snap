// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone Number With Modal should render default props 1`] = `
<div>
  <WithFields(PhoneNumberFields)
    isISOCountryCode={true}
    nameForPhoneCountryCode="countryCode"
    nameForPhoneNumber="phoneNumber"
    names={
      Array [
        "countryCode",
        "phoneNumber",
      ]
    }
    onLabelClick={[Function]}
  />
  <Connect(FullScreenModal)
    id="COUNTRY_CODE"
  >
    <PhoneCountryCodeList
      onCancel={[Function]}
      onCountryCodeSelect={[Function]}
    />
  </Connect(FullScreenModal)>
</div>
`;

exports[`Phone Number With Modal should render non-iso props 1`] = `
<div>
  <WithFields(PhoneNumberFields)
    className="non-iso-codes"
    isISOCountryCode={false}
    nameForPhoneCountryCode="testCaseCountryCode"
    nameForPhoneNumber="testCasePhoneNumber"
    names={
      Array [
        "testCaseCountryCode",
        "testCasePhoneNumber",
      ]
    }
    onLabelClick={[Function]}
  />
  <Connect(FullScreenModal)
    id="TEST_CASE_COUNTRY_CODE"
  >
    <PhoneCountryCodeList
      onCancel={[Function]}
      onCountryCodeSelect={[Function]}
    />
  </Connect(FullScreenModal)>
</div>
`;
